lib/Sema/TypeCheckNameLookup.cpp

```diff
@@ -362,17 +362,22 @@ LookupTypeResult TypeChecker::lookupMemberType(DeclContext *dc,
       }
     }
 
-    // Ignore typealiases found in protocol members.
-    if (auto alias = dyn_cast<TypeAliasDecl>(typeDecl)) {
-      auto aliasParent = alias->getParent();
-      if (aliasParent != dc && isa<ProtocolDecl>(aliasParent))
-        continue;
-    }
-
     // Substitute the base into the member's type.
     if (Type memberType = substMemberTypeWithBase(dc->getParentModule(),
                                                   typeDecl, type,
                                                   /*isTypeReference=*/true)) {
+
+      // Similar to the associated type case, ignore typealiases containing
+      // associated types when looking into a non-protocol.
```

Comment:
It might be clearer to factor out the entire predicate into its own method, because then in addition to comments you can rely on the method name to understand what's going on